Document -XX:+GlobalLockReservation option

IBMJimmyk opened the pull request:
Closes: #535
Signed-off-by: jimmyk <jimmyk@ca.ibm.com>

IBMJimmyk commented:
@gita-omr @SueChaplain  Could you take a look at this? Also is there anyone else I am supposed to tag? This is my first time trying to submit a PR to openj9-docs.

doveye commented:
@IBMJimmyk I've asked @pmhayward to take a look when he's back in tomorrow. Sue's on vacation at the moment.

SueChaplain commented:
Jenkins doc stage

SueChaplain commented on docs/xxgloballockreservation.md:
```diff
@@ -0,0 +1,101 @@
+
+# -XX:\[+|-\]GlobalLockReservation
+
+**(Power only)**
```
Is this AIX and Linux? If so, for consistency, can we say (**AIX and Linux on Power systems**) ... i.e call out the OS as well as the arch.

IBMJimmyk replied:
I'll change that.

SueChaplain commented on docs/xxgloballockreservation.md:
```diff
@@ -0,0 +1,101 @@
+|`-XX:+GlobalLockReservation`| Enable |                                                                                |
+|`-XX:-GlobalLockReservation`| Disable| <i class="fa fa-check" aria-hidden="true"></i><span class="sr-only">yes</span> |
+
+This optimization is targeted towards applications with lots of uncontended locked objects that are being locked just to be safe. When enabled, heuristics are used to try and determine when an object will be exclusively locked by a single thread so that faster, more specialized code can be used for locking the object. Negative performance side effects may occur if the heuristics used get things w ...
```
I think the last sentence here could be re-worded for clarity (I don't think we need to talk about functional correctness - this is implied by the performance statement):

If the heuristics incorrectly identify an object as a target for processing, performance might be adversely affected.

IBMJimmyk replied:
I'll make this change as well.

SueChaplain commented on docs/xxgloballockreservation.md:
```diff
@@ -0,0 +1,101 @@
+
+| Parameter                                                     | Effect                                                                                          |
+|---------------------------------------------------------------|-------------------------------------------------------------------------------------------------|
+| [`reservedTransitionThreshold`](#reservedtransitionthreshold) | Advanced tuning parameter. Changes amount of time spent analyzing an object.                    |
```
Rather than stating *Advanced tuning parameter* for every row in this table, please can we include a lead in sentence before the table:

"Advanced tuning parameters are shown in the following table:"

IBMJimmyk replied:
I can do this as well.

SueChaplain commented on docs/xxgloballockreservation.md:
```diff
@@ -0,0 +1,101 @@
+  |---------------|----------------|-----------------------|
+  | `<value>`     | number         | 1                     |
+
+: Number of times an object is locked by the same thread before it is considered reserved minus 2. So, with the default value of 1 and object can be reserved the third time it is locked. Value can be 0-3 inclusive. Values of 4 or higher are treated as infinity.
```
For clarity, can we say 

"Number of times an object is locked by the same thread before it is considered reserved minus a value of 2. So, with a default value of 1, an object can be reserved the third time it is locked. `<value>` can be 0-3 inclusive. Values of 4 or higher are treated as infinity."

IBMJimmyk replied:
That sounds better so I'll change it.

SueChaplain commented on docs/xxgloballockreservation.md:
```diff
@@ -0,0 +1,101 @@
+  |---------------|----------------|-----------------------|
+  | `<value>`     | number         | 10                    |
+
+: Minimum number of objects of a class that get converted to flat before the class can be considered highly incompatible and objects of that class will never be reserved. Values of 65536 or higher are treated as infinity.
```
Similar comment to last one. Please can we break the long sentence into 2, and remove the future tense "Objects of that class are never reserved."

IBMJimmyk replied:
I'll update this for both the `cancelAbsoluteThreshold` option and `reservedAbsoluteThreshold` option.

SueChaplain commented:
@IBMJimmyk - I staged your PR here: https://eclipse.github.io/openj9-docs-staging/536/xxgloballockreservation/

Well done for getting all the formatting correct - it's always tricky with tables and indented text! 

I made a few review comments for the file in your PR. Please can also update the `mkdocs.yml` file to add this new topic into the table of contents. Please also add a new link and section to the `version0.20.md` file so that this change is in the release notes section.

IBMJimmyk commented:
I rebased my changes since they were too old to see `version0.20.md`.

I made all previously suggested changes to `xxgloballockreservation.md`
I added a new ToC entry to `mkdocs.yml`.

I also added a new section to `version0.20.md`. One issue I had was previous releases notes were very inconsistent on the format of the text when linking to another file. I went with `-XX:[+|-]GlobalLockReservation` but you can see `-XX:[+-]ExitOnOutOfMemoryError` just above it and more variations in previous release note.

The section on `ExitOnOutOfMemoryError` was linking to the wrong location so I changed it to `xxexitonoutofmemoryerror.md`.

SueChaplain commented:
Thanks @IBMJimmyk - I'll look through it all tomorrow. Re. the inconsistent references to -XX options, I have a PR open to fix those too. I'll cancel and/or amend my one so that yours are clear. Thanks for the diligence :-)

SueChaplain commented:
LGTM - Thanks @IBMJimmyk
